### PR TITLE
Add relative numbers for commands by prepending ~

### PR DESCRIPTION
**Pull request by Wuzzy2**
I extended the commands with relative numbers. A relative number is a number in a chat command parameter that is relative to some known absolute value. It is useful to add or subtract something, instead of setting the value directly. The idea is to add support for relative numbers in commands where it makes sense.

From the player's point of view, a number is made relative by prepending the tilde character to the number. For example, instead of `/time 2000`, you can write `/time ~2000`, this means that the value is added to the current time of day instead of setting it. If the time of day was `5000`, this command will result in a time of day of `7000` instead of `2000`.
The `hh:mm` syntax is also supported, just prepend the tilde before the whole time. With the tilde, you can also use negative values for time:

* `/time ~-2:30`: Subtacts 2 hours and 30 minutes from the current time of day.

Another major supported command is `/teleport`. Each coordinate supports relative numbers.

As a shortcut, just entering the tilde without any number is the same as `~0`.

`/teleport ~ ~100 ~` will raise you by 100 nodes, while X and Z stay the same.

The following commands are supported:

* `/teleport`
* `/spawnentity`
* `/deleteblocks`
* `/emergeblocks`
* `/fixlight`
* `/time`

Note that not all commands that contain numbers have relative number support because relative numbers don't always make sense (like for a radius of `/deleteblocks`).

## To do

This PR is a Work in Progress (more or less).

* Figure out how to document this new syntax without massively blowing up the `/help` text
* The code in general feels a little “dirty” to me. Feel free to whine about it, I will understand
* This adds a couple of helper functions, but they are in `core` because I want to use them across files. The helper functions are NOT meant for the official Lua API (so far). Is there a way to mark these functions as “not for Lua API use”?
* Maybe (!) add the helper functions to the Lua API anyway, but easier and more straight-forwards to use, so that games and mods can easily introduce relative numbers as well

## How to test

* `/teleport ~ ~10 ~0`: Raise by 10 nodes
* `/teleport ~10 ~10 ~10`: Add 10 to X, Y and Z of your current position
* `/teleport ~0 ~0 ~-20`: Move -20 on the Z axis
* `/teleport ~ 10 ~`: Teleport to (current\_x, 10, current\_z)
* `/time ~1111`: Advance time by 1111 millihours
* `/time ~2:00`: Advance time by 2 hours
* `/time ~-0:05`: Go back in time by 5 minutes
* `/deleteblocks (~ ~ ~) (~25 ~25 ~25)`: Delete blocks between your current position and [your current position +25 on each axis]
* (also try out the non-relative versions of each command, obviously they must still work perfectly)

## Justification

`/teleport` just got so more convenient as you do not have to memorize your current coords anymore.

### CODE REVIEW

**Comment by appgurueu** [comment]
Suggestion: Don't use `~`, but instead use `+` and `-` for relative numbers

**Comment by sfan5** [comment]
Have you taken a single minute to think about your idea? If so, tell me where `/teleport -10 -10 -10` should take me.

**Comment by Wuzzy2** [comment]
Suggestion rejected.
The minus sign is already taken for absolute negative numbers. `/teleport -20 -20 -20` must still teleport to absolute coordinates (-20, -20, -20).

**Comment by Wuzzy2** [comment]
Update! I have renamed the helper functions, removed a redundant one and added comments to document usage. Maybe they might be even suited for official introduction into `lua_api.txt`, but I am not sure.
The benefit of official introduction is that it will be easier for mods to use the tilde syntax as well. You tell me if it's OK to mark these functions as official part of the Lua API or not.

I also reduced the precision(number of decimal places) of *displayed* coordinates of `/teleport` to 0. Because if you use this command with `~`, you will get dozens of decimal places, making it harder to read.

**Comment by appgurueu** [comment]
> Have you taken a single minute to think about your idea? If so, tell me where /teleport -10 -10 -10 should take me.

I was mainly thinking about time. Also, `~+` chould still be shortened to an explicit `+`.

File: builtin/common/misc_helpers.lua
Comment (appgurueu) [inline builtin/common/misc_helpers.lua:780]: How about
```lua
function core.parse_coordinates(coords, player_name)
	local player = core.get_player_by_name(player_name)
	local absolute_position = player:get_pos()
	if not absolute_position then
		return nil
	end
        local total_coords = {}
        for coord, value in pairs(coords) do
              local total_axis = core.parse_relative_number(value, absolute_position[coord])
              if not total_axis then
                  return nil
               end
        end
	return total_coords
end```

**Comment by Wuzzy2** [comment]
Ugh. I still like the simple tilde notation the best.

Using `+` leads to awkward notation like `/teleport +-10 +-10 +-10` if you want to use the relative value `-10`.

Also, we should definitely use the same notation for all commands, everything else is just pointlessly confusing.

**Comment by appgurueu** [comment]
@Wuzzy2 I did not mean that you should replace the tilde (`~`) with plus (`+`). I was just thinking that `~+3` or `~3` might be better written as `+3`

**Comment by Wuzzy2** [comment]
If you truly believe your alternative is superior, please write out your COMPLETE fully-designed, well-thought out replacement for my concept at once instead of posting random bits of information that create more questions than they answer.

**Comment by ClobberXD** [comment]
Using `+` instead of `~` adds to the confusion, IMO. Tilde is a nice secluded character, that has no special meaning in MT chat. `+` doesn't carry any meaning on its own either, but it's the opposite of `-` which is used to specify negative values. It's also going to be hard to try and explain relative coords to someone when the syntax is as confusing as `/teleport +-5 +10 +-10`. Just my two cents.

**Comment by Wuzzy2** [comment]
Bump after 1 month.

File: builtin/common/misc_helpers.lua
Comment (SmallJoker) [inline builtin/common/misc_helpers.lua:450]: This function is backwards-compatible and adds functionalities modders might also want to make use of.
I think this function name can be kept, and `lua_api.txt` extended to document the additional parameter.

File: builtin/common/misc_helpers.lua
Comment (SmallJoker) [reply builtin/common/misc_helpers.lua:780]: Line count is almost identical, and performance too. For three indices only there's no need to loop.

File: builtin/game/chat.lua
Comment (SmallJoker) [inline builtin/game/chat.lua:880]: I'd prefer to see `hour < 0` checks rather than `negative == "-"` to simplify the regex as much as possible -> `([%d~-]+)`.

**Comment by Wuzzy2** [comment]
@SmallJoker: Done.

**Comment by Wuzzy2** [comment]
@SmallJoker: Done!

**Comment by Wuzzy2** [comment]
Bump after ca. 1.5 months of inactivity.

File: builtin/game/chat.lua
Comment (SmallJoker) [inline builtin/game/chat.lua:936]: What do you need to trim? The expression above does not match spaces.
These lines were added in 70cac9d5 and kept them here for clarification.

**Comment by Wuzzy2** [comment]
Lol, the trim probably isn't needed in that case … Eh, whatever.

**Comment by Wuzzy2** [comment]
What's the progress on this PR?

**Comment by SmallJoker** [comment]
@Wuzzy2 No rebase needed, no changes requested, hence waiting for a second approval.

**Comment by Wuzzy2** [comment]
Since builtin was changed quite significantly in the past, this PR broke. I had to work on it to make it functional in -dev again.

This was a significant rebase/merge which I could not avoid, and sadly it means the old approval doesn't mean much anymore. :-(

This PR needs to be re-reviewed and re-tested.

**Comment by erlehmann** [comment]
I like the proposal. What has to be done to get this approved again?

**Comment by Wuzzy2** [comment]
Testing and reviewing.

**Review by appgurueu — CHANGES_REQUESTED** [review]
Regarding the syntax: Tilde (`~`) seems like a pretty arbitrary choice to me; I know of no other place where it is used to indicate a relative value - usually it means "roughly", or, in Lua, "not". This is how it currently looks:

* Same value: `~0`, `~-0`, `~` as a shortcut
* Increase value: `~+42`, `~42` (the latter feels *very* wrong to me)
* Decrease value: `~-42`

The equals sign (`=`) might be more intuitive, but has a different meaning of, well, equals. Using it as a prefix looks quite odd as well:


* Same value: `=0`, `=-0`, `=` as a shortcut (shortcut is nice, the other two are plain wrong)
* Increase value: `=+42`, `=42`
* Decrease value: `=-42`

of which most are quite confusing.

Finally, the only proper notation I can think of is the one frequently used by programming languages (except Lua):

* Same value: `+=0`, `-=0`, `=` as a shortcut
* Increase value: `+=42`
* Decrease value: `-=42`

The downside of this would be that this slightly complicates pattern matching as opposed to a simple prefix, as the equals sign is inbetween operation/sign and number. As a better alternative, I propose `+` as a prefix:

* Same value: `+0`, `+-0, consider adding `=` as a shortcut
* Increase value: `+42`
* Decrease value: `+-42`

`-42` would keep it's meaning as the absolute value `-42` (required for teleportation etc.)

**Comment by erlehmann** [comment]
> Regarding the syntax: Tilde (~) seems like a pretty arbitrary choice to me; I know of no other place where it is used to indicate a relative value

Minecraft

**Comment by Wuzzy2** [comment]
I have updated the PR and addressed most of the complaints by appgurueu.

I don’t really care *that* much about the character to use, but your arguments against it are just as arbitrary as my character choice. It's a boring discussion. I strongly reject using the `+` because `+-10` is VERY confusing.

> Might be more elegant to handle this in the pattern using another capture group? It would then become (~?)([+-]?%d+).

Nah, I don't care enough about it to change it.

Okay, when I think about it, I think it is okay to include `minetest.parse_relative_number` and `minetest.parse_coordinates` into the Lua API. Why? Because mods like to read numbers as well, so having these parsing helper functions will reduce redundancy and keeps syntax consistent. Without the functions being available for mods, that would limit relative numbers essentially to builtin functions only, which sucks. What do you think?

Finally, I still haven't figured out how to communicate to the user that the tilde syntax is avaliable in the parameter help. I haven't touched it so far. I feel like the chat command help in general might need an extension, a way to write a per-parameter help. Maybe this is something for a future PR, idk. Or does someone have a simple and quick idea for this?

**Comment by pecksin** [comment]
Adding my opinion the tilde ~ is confusing. What about the "at" @ symbol instead? The coord is meant to be relative to where we are currently "at"

¬_¬

**Comment by Wuzzy2** [comment]
@Df458: The rebase is complete! (third rebase now ...)

@pecksin: No bikeshedding, please.

**Review by Df458 — APPROVED** [review]
I'm not sure my test&review this time is up to my usual standard of quality, but looking through the code and testing hasn't really raised any specific issues.

I'm not fond of the extra level of complexity this adds to the relevant chat command code, but it's probably somewhat necessary and it's not something we're likely to touch all the time like core engine code, so... probably ok in this instance?

As far as the current open comments go, I don't see either as a blocker. I'm not bothered by the current exposed methods personally, and while one could argue for adjusting the patterns and such it's a style choice that shouldn't really block a feature like this--it can be resolved in a follow-up PR if anyone feels the need to try and re-organize regardless. On the "which character to use" side of things, ~'s a fine choice _because_ it's not used in basic arithmetic (and so might allow modders to use this standard in equations), as well as the connection to Minecraft. MT may not be MC, but many of the games on this platform share enough DNA that converging on some standards is sensible enough IMO.

Additionally, this PR has been open, looked and, and discussed for nearly 3 years. I think it has received enough time and scrutiny that I'm confident enough to merge this even if my review's not quite as good as usual.

I'll merge this in ~24hrs, barring any very good last-minute reasons not to.

**Review by Df458 — CHANGES_REQUESTED** [review]
Retracting approval to request a couple changes. Special thanks to @erlehmann for the suggestion to play with `nan` and `inf`s in relation to `core.parse_relative_number`...

First of all the aforementioned `core.parse_relative_number` seems quite happy to read `nan`s and `infs`--absolute or relative--and hand them back to the user. I'm not sure this is something we want to support, since the majority of users that call a function like this won't consider such values valid but may not think to explicitly test for them.

Then, this got me thinking about `minetest.string_to_area`... from a quick test, I noticed an interesting bug (relative to 1,1,1 in this test):
`"(1,1,1) (1,1,nan)"` returns nil, but `"(1,1,1) (1,1,~nan)"` returns 1,1,1 and 1,1,1...
...and yet `"(1,1,1) (1,~nan,1)"` returns nil. I'd ask that you do another pass on your publicly-facing API calls and ensure they handle `nan` and `inf` properly. If we look to the current pos-parsing calls as an example, these values should consistently return `nil`

**Comment by rubenwardy** [comment]
The utility functions are the perfect candidate for being unit tested with busted

File: builtin/common/misc_helpers.lua
Comment (erlehmann) [reply builtin/common/misc_helpers.lua:776]: I believe this should be exposed so games can use it. In fact, I would use this in Mineclonia.

**Comment by Zughy** [comment]
@Wuzzy2 are you gonna continue this?

**Comment by Wuzzy2** [comment]
I’m giving up on this one. Code is too old, can no longer compile. Everyone is free to continue the work.

**Comment by Wuzzy2** [comment]
Guess what? I revive this PR. I have fixed the problem with `inf` and `nan` in `minetest.string_to_area`. Those values are now illegal. I also added a couple of new asserts for testing.

**Comment by Df458** [comment]
I will re-review this weekend!

**Review by Df458 — APPROVED** [review]

**Review by Df458 — CHANGES_REQUESTED** [review]
Re-tested, seems fine now with just one exception. I noticed it's possible to make string_to_area parse something like `(1,1,1,1,1,1,1) (-1,2,1,2)` without error... Could be good to prevent it from taking inputs with more than 3 elements

**EDIT** To explain further why I think this should hold up the merge, since I was originally going to let it slide... I feel like we've let not-quite-ok things through in the past and people sometimes end up relying on these behaviors making them difficult to remove later. Since this is a pretty simple one, best to nip it in the bud before it becomes an issue.

**Comment by appgurueu** [comment]
I can make a PR to fix the issue (patterns not being properly anchored because the dirty string splitting requires it)

**Comment by Df458** [comment]
@appgurueu If you're offering to do it, I'm fine approving this now and letting you make a follow-up.

**Review by Df458 — APPROVED** [review]

**Comment by appgurueu** [comment]
> @appgurueu If you're offering to do it, I'm fine approving this now and letting you make a follow-up.

Nono, this is meant as I'm working on a PR to Wuzzy's repo.

**Review by Df458 — CHANGES_REQUESTED** [review]
Rescinding approval then

**Comment by Df458** [comment]
An an aside, the test failure appears to be an actual bug--but not from this PR. We should be able to safely merge this regardless of the red there.

**Comment by Wuzzy2** [comment]
Okay, I'm waiting for the code to drop then.

**Comment by appgurueu** [comment]
> Okay, I'm waiting for the code to drop then.

[Here it is.](https://github.com/Wuzzy2/minetest/pull/3) Make sure to review it thoroughly.

**Comment by Wuzzy2** [comment]
Thanks, @appgurueu, I'm happy with the changes now.

@ other people: Please review. :-)

**Review by Df458 — APPROVED** [review]
LGTM

**Comment by Df458** [comment]
@SmallJoker, given the time delay are you interested in re-reviewing? If you don't respond I'll merge this tomorrow.

**Comment by appgurueu** [comment]
Yay

**Pull request merged**
